docs/content/intro/tutorial-preview-environments.md

```diff
@@ -0,0 +1,433 @@
+as well as auto-deployment of the applications themselves.
+
+Preview Environments are a great place for early experimentation where changes to your 
+application under development can be tested against production version of applications you're depending on.
+
+## Setup
+
+For the purpose of this tutorial we're assuming you have a Kubernetes cluster up and running and Brigade components are deployed to `brigade` namespace within it. We'll be relying on GitHub webhooks so make sure that `brigade-github-app` is also installed.  
+
+[Docker for Desktop's Kubernetes](https://docs.docker.com/docker-for-mac/kubernetes/) 
+cluster is sufficient to perform all steps from this tutorial. To accept incoming GitHub Webhooks to your desktop cluster you can use free version of [Ngrok](https://ngrok.com/) service to establish secure tunnelling. Follow this [excellent guide](https://stefanprodan.com/2018/expose-kubernetes-services-over-http-with-ngrok/) to set this up.
+
+Brigade's [brig](../../install/#brig) cli utility should be present on your machine.
+
+## Git Repositories
+
+We will use two GitHub repositories:
+
+- [brigade-tutorial-config](https://github.com/brigadecore/brigade-tutorial-config): containing orchestration responsible for managing brigade projects and creating new environments.
+- [brigade-tutorial-app](https://github.com/brigadecore/brigade-tutorial-app): example microservice used to demonstrate release process.
+
+`GITHUB_SHARED_SECRET` and `GITHUB_TOKEN` environment variables will be used throughout this tutorial and you'll have to ensure they are set correctly. To generate GitHub token follow [this article](https://help.github.com/en/articles/creating-a-personal-access-token-for-the-command-line).
+
+## Config Brigade Project
+
+Let's create new Brigade Project for our `brigade-tutorial-config` repository.
+
+```sh
+# Ensure brigade charts are added to your helm registry:
+$ helm repo add brigade https://brigadecore.github.io/charts \
+  --kube-context=docker-for-desktop
+
+# Create new Brigade Project by installing brigade-project chart:
+$ helm upgrade brigade-config brigade/brigade-project --install \
+  --namespace brigade \
+  --kube-context docker-for-desktop \
+  --set github.token=$(GITHUB_TOKEN) \
+  --set worker.tag=v1.0.0 \
+  --set project=brigadecore/brigade-tutorial-config \
+  --set repository=github.com/brigadecore/brigade-tutorial-config \
+  --set cloneURL=https://github.com/brigadecore/brigade-tutorial-config
+```
+
+## Create Preview Environment Pipeline
+
+New environments will be created by executing brigade script via `brig` cli. 
+
+![Create Environment Pipeline](https://docs.brigade.sh/img/preview-environments-create.png)
+
+## Create Namespace
+
+Let's start by adding new `brigade.js` to the root of your repository and adding script 
+that will handle creation of new Kubernetes namespace.
+
+```js
+
+// brigade.js
```

Comment:
Our pipelines don't usually deal with application logic, but rather with the pipeline itself. An alternative to using the Kubernetes Node client in `brigade.js` would be to create a job that uses a container with `kubectl` and execute the commands there.

I don't really have a strong preference, and I'd defer this decision to you, but this approach might fit better with the overall direction we point users to take with Brigade pipelines.

What do you think?

Reply:
While obviously I took a different approach here, for the good of the project I think having consistent message for the user would be more desired. My whole flow at the moment heavily relies on k8s Node client and I would have to see if everything is possible (I'm sure it is) by just using kubectl containers running as a job. Using Node client simplifies tasks like checking for existing resources or verifying of some resources are ready. Wonder how easy that sort of coordination will be when results of running kubectl in the container will have to be evaluated by pipeline... I'll investigate...

Reply:
At the same time, I wouldn't want to keep a first iteration of this from getting merged - maybe adding a note suggesting the alternative approach might be enough while we're investigating.